File: app/src/main/java/com/zack/enderplan/view/activity/HomeActivity.java
package com.zack.enderplan.view.activity;

import android.animation.ObjectAnimator;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.design.widget.FloatingActionButton;
import android.support.design.widget.NavigationView;
import android.support.v4.view.GravityCompat;
import android.support.v4.widget.DrawerLayout;
import android.support.v7.app.ActionBarDrawerToggle;
import android.support.v7.widget.Toolbar;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.TextView;
import android.widget.Toast;

import com.zack.enderplan.App;
import com.zack.enderplan.R;
import com.zack.enderplan.common.Util;
import com.zack.enderplan.injector.component.DaggerHomeComponent;
import com.zack.enderplan.injector.module.HomePresenterModule;
import com.zack.enderplan.view.fragment.BaseListFragment;
import com.zack.enderplan.view.fragment.MyPlansFragment;
import com.zack.enderplan.view.fragment.AllTypesFragment;
import com.zack.enderplan.presenter.HomePresenter;
import com.zack.enderplan.view.contract.HomeViewContract;
import com.zack.enderplan.common.Constant;

import javax.inject.Inject;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;

public class HomeActivity extends BaseActivity implements HomeViewContract {

    @BindView(R.id.toolbar)
    Toolbar mToolbar;
    @BindView(R.id.fab_create)
    FloatingActionButton mCreateFab;
    @BindView(R.id.navigator)
    NavigationView mNavigator;
    @BindView(R.id.layout_drawer)
    DrawerLayout mDrawerLayout;

    @Inject
    HomePresenter mHomePresenter;

    private TextView mUcPlanCountText;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        mHomePresenter.attach();
    }

    @Override
    protected void onInjectPresenter() {
        DaggerHomeComponent.builder()
                .homePresenterModule(new HomePresenterModule(this))
                .appComponent(App.getAppComponent())
                .build()
                .inject(this);
    }

    @Override
    protected void onPostCreate(Bundle savedInstanceState) {
        super.onPostCreate(savedInstanceState);
        mHomePresenter.notifyStartingUpCompleted();
        //如果放到setInitialView中，toolbar的title不会改变
        showFragment(savedInstanceState == null ? Constant.MY_PLANS : savedInstanceState.getString(Constant.CURRENT_FRAGMENT, Constant.MY_PLANS));
    }

    @Override
    protected void onSaveInstanceState(Bundle outState) {
        super.onSaveInstanceState(outState);
        String currentFragment;
        if (isFragmentShowing(Constant.MY_PLANS)) {
            currentFragment = Constant.MY_PLANS;
        } else {
            currentFragment = Constant.ALL_TYPES;
        }
        outState.putString(Constant.CURRENT_FRAGMENT, currentFragment);
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        mHomePresenter.detach();
    }

    @Override
    public void onBackPressed() {
        mHomePresenter.notifyBackPressed(
                mDrawerLayout.isDrawerOpen(GravityCompat.START),
                isFragmentShowing(Constant.MY_PLANS)
        );
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        // Inflate the menu; this adds items to the action bar if it is present.
        getMenuInflater().inflate(R.menu.menu_home, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        // Handle action bar item clicks here. The action bar will
        // automatically handle clicks on the Home/Up button, so long
        // as you specify a parent activity in AndroidManifest.xml.
        switch (item.getItemId()) {
            case R.id.action_settings:
                SettingsActivity.start(this);
                break;
            case R.id.action_about:
                AboutActivity.start(this);
                break;
            default:
                break;
        }
        return super.onOptionsItemSelected(item);
    }

    @Override
    public void showInitialView(String ucPlanCount) {
        setContentView(R.layout.activity_home);
        ButterKnife.bind(this);

        mUcPlanCountText = ButterKnife.findById(mNavigator.getHeaderView(0), R.id.text_uc_plan_count);

        setSupportActionBar(mToolbar);

        ActionBarDrawerToggle toggle = new ActionBarDrawerToggle(this, mDrawerLayout, mToolbar, R.string.navigation_drawer_open, R.string.navigation_drawer_close);
        mDrawerLayout.addDrawerListener(toggle);
        toggle.syncState();

        mNavigator.setNavigationItemSelectedListener(new NavigationView.OnNavigationItemSelectedListener() {
            @Override
            public boolean onNavigationItemSelected(@NonNull MenuItem item) {
                switch (item.getItemId()) {
                    case R.id.nav_my_plans:
                        showFragment(Constant.MY_PLANS);
                        break;
                    case R.id.nav_all_types:
                        showFragment(Constant.ALL_TYPES);
                        break;
                    case R.id.nav_settings:
                        enterActivity(Constant.SETTINGS);
                        break;
                    case R.id.nav_about:
                        enterActivity(Constant.ABOUT);
                        break;
                    default:
                        break;
                }
                mDrawerLayout.closeDrawer(GravityCompat.START);
                return true;
            }
        });

        mUcPlanCountText.setText(ucPlanCount);
        mUcPlanCountText.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                mHomePresenter.notifyUcPlanCountTextClicked();
            }
        });
    }

    @Override
    public void changeUcPlanCount(String ucPlanCount) {
        mUcPlanCountText.setText(ucPlanCount);
    }

    @Override
    public void closeDrawer() {
        mDrawerLayout.closeDrawer(GravityCompat.START);
    }

    @Override
    public void changeFabVisibility(boolean isVisible) {
        ObjectAnimator.ofFloat(mCreateFab, "translationY", mCreateFab.getTranslationY(), isVisible ? 0f : Util.convertDpToPx(Constant.FAB_COORDINATE) + mCreateFab.getHeight() / 2f)
                .setDuration(200)
                .start();
    }

    @Override
    public void showFragment(String tag) {
        if (!isFragmentShowing(tag)) {
            BaseListFragment fragment;
            switch (tag) {
                case Constant.MY_PLANS:
                    fragment = new MyPlansFragment();
                    break;
                case Constant.ALL_TYPES:
                    fragment = new AllTypesFragment();
                    break;
                default:
                    throw new IllegalArgumentException("The argument tag cannot be " + tag);
            }
            fragment.setOnListScrolledListener(new BaseListFragment.OnListScrolledListener() {
                @Override
                public void onListScrolled(int variation) {
                    mHomePresenter.notifyListScrolled(variation);
                }
            });
            getSupportFragmentManager().beginTransaction().replace(R.id.layout_fragment, fragment, tag).commit();
        }
        //在切换相同fragment时，下面的语句是不必要的
        int titleResId, fabResId, navViewCheckedItemId;
        switch (tag) {
            case Constant.MY_PLANS:
                titleResId = R.string.title_fragment_my_plans;
                fabResId = R.drawable.ic_add_black_24dp;
                navViewCheckedItemId = R.id.nav_my_plans;
                break;
            case Constant.ALL_TYPES:
                titleResId = R.string.title_fragment_all_types;
                fabResId = R.drawable.ic_playlist_add_black_24dp;
                navViewCheckedItemId = R.id.nav_all_types;
                break;
            default:
                throw new IllegalArgumentException("The argument tag cannot be " + tag);
        }
        mToolbar.setTitle(titleResId);
        mCreateFab.setImageResource(fabResId);
        mCreateFab.setTranslationY(0f);
        mNavigator.setCheckedItem(navViewCheckedItemId);
    }

    @Override
    public void onPressBackKey() {
        super.onBackPressed();
    }

    @Override
    public void enterActivity(String tag) {
        switch (tag) {
            case Constant.GUIDE:
                GuideActivity.start(this);
                break;
            case Constant.SETTINGS:
                SettingsActivity.start(this);
                break;
            case Constant.ABOUT:
                AboutActivity.start(this);
                break;
        }
    }

    @Override
    public void showToast(String msg) {
        Toast.makeText(this, msg, Toast.LENGTH_SHORT).show();
    }

    @OnClick({R.id.fab_create})
    public void onClick(View view) {
        switch (view.getId()) {
            case R.id.fab_create:
                if (isFragmentShowing(Constant.MY_PLANS)) {
                    PlanCreationActivity.start(this);
                } else {
                    TypeCreationActivity.start(this);
                }
                break;
        }
    }

    private boolean isFragmentShowing(String tag) {
        return getSupportFragmentManager().findFragmentByTag(tag) != null;
    }
}

File: app/src/main/java/net/zackzhang/app/end/view/activity/HomeActivity.kt
package net.zackzhang.app.end.view.activity

import android.content.Context
import android.content.Intent
import android.graphics.Color
import android.os.Bundle
import android.support.v4.app.Fragment
import android.support.v4.app.FragmentTransaction
import android.support.v4.view.GravityCompat
import android.support.v7.app.ActionBarDrawerToggle
import android.view.Menu
import android.view.MenuItem
import android.widget.TextView
import android.widget.Toast
import kotlinx.android.synthetic.main.activity_home.*
import net.zackzhang.app.end.App
import net.zackzhang.app.end.R
import net.zackzhang.app.end.common.Constant
import net.zackzhang.app.end.injector.component.DaggerHomeComponent
import net.zackzhang.app.end.injector.module.HomePresenterModule
import net.zackzhang.app.end.presenter.HomePresenter
import net.zackzhang.app.end.view.contract.HomeViewContract
import net.zackzhang.app.end.view.fragment.AllTypesFragment
import net.zackzhang.app.end.view.fragment.MyPlansFragment
import javax.inject.Inject

class HomeActivity : BaseActivity(), HomeViewContract {

    companion object {

        fun start(context: Context) {
            context.startActivity(Intent(context, HomeActivity::class.java).setFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP))
        }
    }

    @Inject
    lateinit var homePresenter: HomePresenter

    // 不使用synthetic，因为不会从cache中获取子view
    private val planCountText by lazy { navigator.getHeaderView(0).findViewById<TextView>(R.id.text_plan_count) }
    private val planCountDscptText by lazy { navigator.getHeaderView(0).findViewById<TextView>(R.id.text_plan_count_dscpt) }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        homePresenter.attach()
    }

    override fun onInjectPresenter() {
        DaggerHomeComponent.builder()
                .homePresenterModule(HomePresenterModule(this))
                .appComponent(App.appComponent)
                .build()
                .inject(this)
    }

    override fun onRestoreInstanceState(savedInstanceState: Bundle) {
        super.onRestoreInstanceState(savedInstanceState)
        homePresenter.notifyInstanceStateRestored(savedInstanceState.getString(Constant.CURRENT_FRAGMENT, Constant.MY_PLANS))
    }

    override fun onPostCreate(savedInstanceState: Bundle?) {
        super.onPostCreate(savedInstanceState)
        homePresenter.notifyStartingUpCompleted()
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        homePresenter.notifySavingInstanceState(outState)
    }

    override fun onDestroy() {
        super.onDestroy()
        homePresenter.detach()
    }

    override fun onBackPressed() {
        homePresenter.notifyBackPressed(
                layout_drawer.isDrawerOpen(GravityCompat.START),
                isFragmentShowing(Constant.MY_PLANS)
        )
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.menu_home, menu)
        menu.findItem(R.id.action_search).icon.setTint(Color.WHITE)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.action_search -> homePresenter.notifySearchButtonTouched()
        }
        return super.onOptionsItemSelected(item)
    }

    // 参数data必须声明成可选类型，否则崩溃，因为若没有数据返回，data为null
    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        when (resultCode) {
            //正常结束引导
            RESULT_OK -> { }
            //中途退出引导
            RESULT_CANCELED -> exit()
        }
    }

    override fun showInitialView(planCount: String, textSize: Int, planCountDscpt: String) {
        setContentView(R.layout.activity_home)

        setSupportActionBar(toolbar)

        val toggle = ActionBarDrawerToggle(this, layout_drawer, toolbar, R.string.navigation_drawer_open, R.string.navigation_drawer_close)
        layout_drawer.addDrawerListener(toggle)
        toggle.syncState()

        navigator.setNavigationItemSelectedListener {
            when (it.itemId) {
                R.id.nav_my_plans -> homePresenter.notifySwitchingFragment(Constant.MY_PLANS)
                R.id.nav_all_types -> homePresenter.notifySwitchingFragment(Constant.ALL_TYPES)
                R.id.nav_settings -> startActivity(Constant.SETTINGS)
                R.id.nav_about -> startActivity(Constant.ABOUT)
            }
            layout_drawer.closeDrawer(GravityCompat.START)
            true
        }

        fab_create.setOnClickListener { homePresenter.notifyCreationButtonTouched() }

        changeDrawerHeaderDisplay(planCount, textSize, planCountDscpt)
    }

    override fun showInitialFragment(restored: Boolean, shownTag: String) {
        val transaction = supportFragmentManager.beginTransaction()
        if (restored) {
            supportFragmentManager.fragments
                    .map { it.tag }
                    // 不隐藏 tag 为 null 的 fragment，因为这是 WeatherFragment 的子页（每个城市）
                    .filter { it != null && it != shownTag }
                    .forEach { it?.let { transaction.hide(it) } }
        } else {
            // 如果 activity 正常启动，仅添加 shownTag 对应的 fragment
            transaction.add(shownTag)
        }
        transaction.commit()

        updateAfterUpdatingFragment(shownTag)
    }

    override fun switchFragment(fromTag: String, toTag: String) {
        if (!isFragmentExist(toTag)) {
            // 如果要显示的 fragment 还未创建，创建
            supportFragmentManager.beginTransaction().hide(fromTag).add(toTag).commit()
        } else if (!isFragmentShowing(toTag)) {
            // 如果要显示的 fragment 已创建但未显示，显示
            supportFragmentManager.beginTransaction().hide(fromTag).show(toTag).commit()
        }
        // 如果要显示的 fragment 已创建且已显示，不做任何操作

        updateAfterUpdatingFragment(toTag)
    }

    override fun changePlanCount(planCount: String, textSize: Int) {
        planCountText.text = planCount
        planCountText.textSize = textSize.toFloat()
    }

    override fun changeDrawerHeaderDisplay(planCount: String, textSize: Int, planCountDscpt: String) {
        changePlanCount(planCount, textSize)
        planCountDscptText.text = planCountDscpt
    }

    override fun closeDrawer() {
        layout_drawer.closeDrawer(GravityCompat.START)
    }

    override fun onPressBackKey() {
        super.onBackPressed()
    }

    override fun startActivity(tag: String) {
        when (tag) {
            Constant.GUIDE -> GuideActivity.start(this)
            Constant.PLAN_SEARCH -> PlanSearchActivity.start(this)
            Constant.TYPE_SEARCH -> TypeSearchActivity.start(this)
            Constant.SETTINGS -> SettingsActivity.start(this)
            Constant.ABOUT -> AboutActivity.start(this)
            Constant.PLAN_CREATION -> PlanCreationActivity.start(this)
            Constant.TYPE_CREATION -> TypeCreationActivity.start(this)
        }
    }

    override fun showToast(msg: String) {
        Toast.makeText(this, msg, Toast.LENGTH_SHORT).show()
    }

    private fun updateAfterUpdatingFragment(tag: String) {
        val titleResId: Int
        val checkedItemId: Int
        when (tag) {
            Constant.MY_PLANS -> {
                titleResId = R.string.title_fragment_my_plans
                checkedItemId = R.id.nav_my_plans
            }
            Constant.ALL_TYPES -> {
                titleResId = R.string.title_fragment_all_types
                checkedItemId = R.id.nav_all_types
            }
            else -> throw IllegalArgumentException("The argument tag cannot be \"$tag\"")
        }
        toolbar.setTitle(titleResId)
        fab_create.translationY = 0f
        navigator.setCheckedItem(checkedItemId)
    }

    private fun findFragment(tag: String) = supportFragmentManager.findFragmentByTag(tag)

    private fun isFragmentExist(tag: String) = findFragment(tag) != null

    private fun isFragmentShowing(tag: String): Boolean {
        val fragment = findFragment(tag)
        return fragment != null && !fragment.isHidden
    }

    private fun FragmentTransaction.add(tag: String): FragmentTransaction {
        add(
                R.id.layout_fragment,
                when (tag) {
                    Constant.MY_PLANS -> MyPlansFragment()
                    Constant.ALL_TYPES -> AllTypesFragment()
                    // 在这里添加新 fragment
                    else -> throw IllegalArgumentException("The argument tag cannot be \"$tag\"")
                },
                tag
        )
        return this
    }

    private fun FragmentTransaction.hide(tag: String): FragmentTransaction {
        hide(findFragment(tag) ?: throw IllegalArgumentException("No Fragment with tag \"$tag\" found"))
        return this
    }

    private fun FragmentTransaction.show(tag: String): FragmentTransaction {
        show(findFragment(tag) ?: throw IllegalArgumentException("No Fragment with tag \"$tag\" found"))
        return this
    }
}
